app/debrief_svc.py

```diff
@@ -130,6 +130,27 @@ async def _build_prop_d3(self, operation_ids, prop):
                     previous_prop_graph_id = prop_graph_id
         return graph_output
 
+    @staticmethod
+    def generate_ttps(operations):
+        ttps = dict()
+        for op in operations:
+            for link in op.chain:
+                if not link.cleanup:
+                    tactic_name = link.ability.tactic
+                    if tactic_name not in ttps.keys():
+                        tactic = dict(name=tactic_name,
+                                      techniques={link.ability.technique_name: link.ability.technique_id},
+                                      steps={op.name: [link.ability.name]})
+                        ttps[tactic_name] = tactic
+                    else:
+                        if link.ability.technique_name not in ttps[tactic_name]['techniques'].keys():
+                            ttps[tactic_name]['techniques'][link.ability.technique_name] = link.ability.technique_id
+                        if op.name not in ttps[tactic_name]['steps'].keys():
+                            ttps[tactic_name]['steps'][op.name] = [link.ability.name]
+                        elif link.ability.name not in ttps[tactic_name]['steps'][op.name]:
+                            ttps[tactic_name]['steps'][op.name].append(link.ability.name)
+        return dict(sorted(ttps.items()))
```

Comment:
I think by breaking this up we could increase it readability. As it stand its fairly complex.

Reply:
Made some helper functions to break it up

Reply:
small change but way better to read